File: app/src/main/java/com/waracle/androidtest/MainActivity.java
package com.waracle.androidtest;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import com.waracle.androidtest.Adapters.MainAdapter;
import com.waracle.androidtest.Model.CakeModel;
import com.waracle.androidtest.databinding.ActivityMainBinding;
import com.waracle.androidtest.viewModels.CakeViewModel;

import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.RecyclerView;


public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;

    private CakeViewModel cakeViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        cakeViewModel = ViewModelProviders.of(this).get(CakeViewModel.class);
        cakeViewModel.getCakeModels().observe(this, new Observer<List<CakeModel>>() {
            @Override
            public void onChanged(List<CakeModel> cakeModels) {
                if (recyclerView.getAdapter() != null) {
                    ((MainAdapter) recyclerView.getAdapter()).setItems(cakeModels);
                }
            }
        });

        ActivityMainBinding activityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        activityMainBinding.setCakeViewModel(cakeViewModel);
        recyclerView = findViewById(R.id.cake_list);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    public void refreshList() {
        if (recyclerView.getAdapter() != null) {
            ((MainAdapter) recyclerView.getAdapter()).clear();
        }
        MainApplication.getDataSource().clearDataSources();
        cakeViewModel.fetchCakeModels();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_refresh) {
            refreshList();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/waracle/androidtest/MainActivity.kt
package com.waracle.androidtest

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.waracle.androidtest.adapters.MainAdapter
import com.waracle.androidtest.model.CakeModel
import com.waracle.androidtest.databinding.ActivityMainBinding
import com.waracle.androidtest.viewModels.CakeViewModel
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity: AppCompatActivity() {

    lateinit var cakeViewModel :CakeViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        cakeViewModel = ViewModelProviders.of(this).get(CakeViewModel::class.java)
        val activityMainBinding:ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        cakeViewModel.cakeModels.observe(this, Observer<MutableList<CakeModel>> {
            if(cake_list.adapter != null) {
                (cake_list.adapter as MainAdapter).setItems(it)
            }
        })
        activityMainBinding.cakeViewModel = cakeViewModel
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_main,menu)
        return true
    }

    fun refreshList(){
        (cake_list.adapter as MainAdapter).clear()
        cakeViewModel.fetchCakeModels()
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when(item!!.itemId){
            R.id.action_refresh ->  {
                refreshList()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}